lib/wallet/coinselector.js

```diff
@@ -0,0 +1,882 @@
+    // try to find a changeless solution using BnB
+    let result = this.selectBnB(target, costOfChange);
+    if (result && result.length > 0) {
+      bestSelection = result;
+      bestWaste = this.getWaste(result, costOfChange, target);
+    }
+
+    // fallback to other algorithms as we can't find a changeless solution
+    // update target to include cost of producing change
+    target += this.getFee(changeOutputSize);
+
+    // find a solution of target using Lowest Larger
+    result = this.selectLowestLarger(target);
+    if (result.length > 0) {
+      const currWaste = this.getWaste(result, costOfChange, target);
+      if (currWaste < bestWaste) {
+        bestSelection = result;
+        bestWaste = currWaste;
+        isChange = true;
+      }
+    }
+
+    // find a solution of target using SRD
+    result = this.selectSRD(target);
+    if (result.length > 0) {
+      const currWaste = this.getWaste(result, costOfChange, target);
+      if (currWaste <= bestWaste) {
+        bestSelection = result;
+        bestWaste = currWaste;
+        isChange = true;
+      }
+    }
+
+    if (bestSelection && bestSelection.length > 0) {
+      for (const i of bestSelection) {
+        const coin = this.coins[this.coinPointers[i].index];
+        this.tx.addCoin(coin);
+        this.chosen.push(coin);
+      }
+      // add change output if solution is not found by BnB
+      if (isChange)
+        this.tx.outputs.push(change);
+      // update fee
+      size = await this.tx.estimateSize(this.getAccount);
+      this.fee = this.getFee(size);
+    }
+  }
+
+  /**
+   * Initiate selection using Single Random Draw selection.
+   * @param {Number} target - Selection target
+   * @returns {Number[]} selected - array of indicies of selected coins
+   */
+
+  selectSRD(target) {
+    // create an array of indices and randomly sort it
+    const arr = [...this.coinPointers.keys()];
+    arr.sort(() => Math.random() - 0.5);
+
+    let index = 0;
+    const selected = [];
+    let selectedValue = 0;
+
+    while (index < arr.length) {
+      const pointer = this.coinPointers[arr[index]];
+      selectedValue += pointer.effectiveValue;
+      selected.push(arr[index]);
+      if (selectedValue >= target)
+        return selected;
+      index++;
+    }
+
+    return [];
+  }
+
+  /**
+   * Initiate selection using Branch and Bound selection.
+   * @param {Number} target - Selection target
+   * @param {Number} costOfChange - Cost of producing and spending change
+   * @returns {Number[]} selected - array of indicies of selected coins
+   */
+
+  selectBnB(target, costOfChange) {
+    const selected = [];
+    let currValue = 0;
+    let remainingValue = 0;
+    let index = 0;
+    for (const pointer of this.coinPointers)
+      remainingValue += pointer.effectiveValue;
+
+    // we don't have enough funds
+    if (remainingValue < target) {
+      throw new FundingError(
+        'Not enough funds.',
+        remainingValue,
+        target);
+    }
```

Comment:
Do this in `init()` instead? Basically, abort right away if we know we can't afford the TX.